src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,471 @@
+            if (localKeys.size() == 1) {
+                // return this element
+                return localKeys.get(0);
+            } else {
+                byte[] bb = new byte[0];
+                for (SecretKey workItem : localKeys) {
+                    byte[] workItemBytes = CipherCore.getKeyBytes(workItem);
+
+                    bb = Arrays.copyOf(bb, bb.length + workItemBytes.length);
+                    System.arraycopy(workItemBytes, 0, bb, bb.length - workItemBytes.length,
+                                     workItemBytes.length);
+                }
+                return new SecretKeySpec(bb, "RAW");
+            }
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Perform the HMAC-Extract operation.
+     *
+     * @param inputKey
+     *     the input keying material used for the HKDF-Extract operation.
+     * @param salt
+     *     the salt value used for HKDF-Extract.  If no salt is to be used a {@code null} value
+     *     should be provided.
+     *
+     * @return a byte array containing the pseudorandom key (PRK)
+     *
+     * @throws InvalidKeyException
+     *     if an invalid salt was provided through the {@code HkdfParameterSpec}
+     */
+    protected byte[] hkdfExtract(SecretKey inputKey, byte[] salt) throws InvalidKeyException {
+
+        if (salt == null) {
+            salt = new byte[hmacLen];
+        }
+        hmacObj.init(new SecretKeySpec(salt, "HKDF-Salt"));
+
+        return hmacObj.doFinal(inputKey.getEncoded());
+    }
+
+    /**
+     * Perform the HMAC-Expand operation.  At the end of the operation, the keyStream instance
+     * variable will contain the complete KDF output based on the input values and desired length.
+     *
+     * @param prk
+     *     the pseudorandom key used for HKDF-Expand
+     * @param info
+     *     optional context and application specific information or {@code null} if no info data is
+     *     provided.
+     * @param outLen
+     *     the length in bytes of the required output
+     *
+     * @return a byte array containing the complete KDF output.  This will be at least as long as
+     * the requested length in the {@code outLen} parameter, but will be rounded up to the nearest
+     * multiple of the HMAC output length.
+     *
+     * @throws InvalidKeyException
+     *     if an invalid key was provided through the {@code HkdfParameterSpec} or derived during
+     *     the generation of the PRK.
+     */
+    protected byte[] hkdfExpand(SecretKey prk, byte[] info, int outLen) throws InvalidKeyException {
+        byte[] kdfOutput;
+
+        // Calculate the number of rounds of HMAC that are needed to
+        // meet the requested data.  Then set up the buffers we will need.
+        hmacObj.init(prk);
+        if (info == null) {
+            info = new byte[0];
+        }
+        int rounds = (outLen + hmacLen - 1) / hmacLen;
+        kdfOutput = new byte[rounds * hmacLen];
```

Comment:
Are we missing a check to ensure that the `outLen` parameter is less than or equal to 255 * HashLen?  See RFC 5869 sec. 2.3.